src/tss2-tcti/tcti-tbs.c

```diff
@@ -0,0 +1,286 @@
+/* SPDX-License-Identifier: BSD-2 */
+/*
+ * Copyright (c) 2015 - 2018 Intel Corporation
+ * All rights reserved.
+ */
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <windows.h>
+#include <Tbs.h>
+
+#include "tss2_tcti.h"
+#include "tss2_tcti_tbs.h"
+
+#include "tcti-common.h"
+#include "tcti-tbs.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * This function wraps the "up-cast" of the opaque TCTI context type to the
+ * type for the TBS TCTI context. The only safe-guard we have to ensure
+ * this operation is possible is the magic number for the TBS TCTI context.
+ * If passed a NULL context, or the magic number check fails, this function
+ * will return NULL.
+ */
+TSS2_TCTI_TBS_CONTEXT*
+tcti_tbs_context_cast (TSS2_TCTI_CONTEXT *tcti_ctx)
+{
+    if (tcti_ctx != NULL && TSS2_TCTI_MAGIC (tcti_ctx) == TCTI_TBS_MAGIC) {
+        return (TSS2_TCTI_TBS_CONTEXT*)tcti_ctx;
+    }
+    return NULL;
+}
+/*
+ * This function down-casts the TBS TCTI context to the common context
+ * defined in the tcti-common module.
+ */
+TSS2_TCTI_COMMON_CONTEXT*
+tcti_tbs_down_cast (TSS2_TCTI_TBS_CONTEXT *tcti_tbs)
+{
+    if (tcti_tbs == NULL) {
+        return NULL;
+    }
+    return &tcti_tbs->common;
+}
+
+TSS2_RC
+tcti_tbs_transmit (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t command_size,
+    const uint8_t *command_buffer)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+    TBS_RESULT tbs_rc;
+    /* Reset resSize */
+    tcti_tbs->resSize = TBS_RESULT_MAX_BUFFER_SIZE;
+
+    if (tcti_tbs == NULL) {
```

Comment:
Still sorting out y'alls coding standards, so forgive me if this doesn't make sense/is a bad comment, but could this check perhaps be pushed up to after line 57? In the event `tcti_tbs == NULL` you could bail out early and avoid the call to `tcti_tbs_down_cast` as well as resource declarations.